### PR TITLE
Proposal: Dynamic Coupling via Local Build Context

### PR DESCRIPTION
**Pull request by WhisperingChaos**
The proposal contained within DynamicCouplingLocalContext.md has been repeated below.
### TLDR

Constructs a virtual build context derived from the uploaded one, similar to creating a directory structure comprised entirely of symbolic links, that exactly specifies what's needed or exposed to a FROM statement and its related Dockerfile commands. This supports writing a series of ADD/COPY commands according to a given interface (context local to the FROM) so they can be decoupled from the uploaded build context.  Decoupling encourages reuse, for example, encapsulating ADD/COPY commands within ONBUILD triggers, as their source path/file references can be redirected to the appropriate ones for any uploaded build context without having to rewrite them.
### TOC

<a href="#Description">Description</a>
<a href="#Syntax">Syntax</a>
<a href="#Semantics">Semantics</a>
<a href="#Benefits">Benefits</a>
<a href="#Example">Example</a>
### Description  <a name="Description">

Dockerfile commands pathologically couple to the structure of the build context provided by "docker build" and since the structure of a build context varies markedly from one "docker build" image request to the next, certain Dockerfile binding commands, like ADD, must be "repeated": manually encoded to reflect the unique structure of each image's build context.  Moreover, this harmful coupling diminishes the utility of both present and future language elements that would benefit from the ability to statically bind to a path/file name (that acts like a variable name) when writing Dockerfile commands, that when executed, are then dynamically resolved and (dynamically) coupled to the actual path/file instance (variable value).  For example, the ADD commands of an image's [ONBUILD triggers](https://docs.docker.com/reference/builder/#onbuild) must either specify path/file names which also requires these names be mirrored by all build contexts which seek to inherit it, or utilize a static coding form that dynamically extends itself, like "ADD . /src", in which '.' immediately adapts to include all resources defined by the build context.  Neither choice represents a long-term solution.

The first method not only requires individuals using an image to adhere to a dictated build context, it additionally burdens those developing images to coordinate source path/file references, using either a centralize or distributed means, to avoid potential overlapping ones.  The second option, the one which dynamically extents the scope of an image's consumption, represents a vulnerability that can be maliciously or unintentionally exploited.  It also diminishes the reliability of processes executing within an <a href="#TermImageContext">_ImageContext_</a>, as additional unintended artifacts (resources) are copied into the destination specified by the command.  Review this [post](https://github.com/docker/docker/issues/7149#issuecomment-84171831) for a more detailed discussion regarding this vulnerability.

This proposal introduces a build context mapping mechanism that relates the static path/file references, encoded on Dockerfile commands required to execute a process(es), like a go compile, within an <a href="#TermImageContext">_ImageContext_</a>, to the rapidly growing set of Dockerfile build contexts generated by developers leveraging this process(es).  Essentially, the proposal provides a means to assemble a <a name="TermLocalBuildContext"></a>_LocalBuildContext_: a tailored version of the build context for a process(es) running within an <a href="#TermImageContext">_ImageContext_</a>.  As suggested by its intent, the mapping mechanism associates a build context reference, like a path or file, to a corresponding one encoded within a Dockerfile command.  To record these mappings, this proposal recommends extending the current FROM Dockerfile command with a new keyword called "CONTEXT". This purposeful association of "CONTEXT" to FROM stems from:
- The current semantics of ONBUILD triggers, as they stipulate their execution to occur within the lexical encoding of FROM, currently preventing their override or injection of other Dockerfile commands ([#8240](https://github.com/docker/docker/issues/8240) would permit injection).  More importantly, CONTEXT's encoded mappings are generally unique to a given build context requiring frequent encoding of "CONTEXT".  If CONTEXT were encoded as an individual Dockerfile command, trigger injection would become the norm instead of the exception, undermining a trigger's encapsulation quality. 
- Its analogous behavior, when containing ONBUILD triggers and/or a sequence of associated Dockerfile commands, like RUN, to ["docker run"](https://docs.docker.com/reference/run/) which in turn resembles a [function's](http://en.wikipedia.org/wiki/Subroutine) invocation.  In a sense, "CONTEXT" defines the parameter list being passed to the <a href="#TermImageContext">_ImageContext_</a> initiated by the FROM keyword.

The remainder of the proposal's presentation discusses CONTEXT's <a href="#Syntax">Syntax</a>, <a href="#Semantics">Semantics</a>, <a href="#Benefits">Benefits</a>, and provides an <a href="#Example">Example</a> to demonstrate its utility.
### Syntax <a name="Syntax"></a>

```
    FROM ... [CONTEXT {\<BuildContextReference\>:\<BuildContextForImageReference\>
                      [\<BuildContextReference\>:\<BuildContextForImageReference\>]... | NULL}]
```
#### Terms
- [_FROM ..._](https://docs.docker.com/reference/builder/#from)
- _<BuildContextReference>_<a name="TermsBuildContextReference">: Defines a path/file reference resolved within the build context provided by "docker build". The reference may contain wildcards conforming to the syntax and semantics of [Go's filepath.Match,](http://golang.org/pkg/path/filepath#Match)
- _<BuildContextForImageReference>_<a name="TermsBuildContextForImageReference">: A path or file reference to appear in the <a href="#TermLocalBuildContext">_LocalBuildContext_</a> assembled for the <a href="#TermImageContext">_ImageContext_</a> initiated by FROM.  This assembled context conforms to the interface expected by or/and defines the interface exposed to an _ImageContext_ when executing its Dockerfie commands.
- _ContextMapEntry_: A paired specification of <BuildContextReference>:<BuildContextForImageReference>.  It defines two concepts:
  - _ContributingSet_ : The set of artifacts (path/file) objects supplied by the build context.  Resolution of _<BuildContextReference>_ determines the members of this set.
  - _ContributingSetLabel_ : A name assigned to a _ContributingSet_.  This name appears in the _LocalBuildContext_ as referencing the _ContributingSet_.  _ContributingSetLabel_ is assigned the value of <BuildContextForImageReference>.
- <a name="TermImageContext"></a>_ImageContext_: A process execution environment constructed by an initiating FROM and its associated Dockerfie commands.   A Dockerfile command associates/couples to the lexically nearest FROM statement that precedes it.
### Semantics <a name="Semantics"></a>

Conceptual method to assemble the <a href="#TermLocalBuildContext">_LocalBuildContext_</a>.
- For each _ContextMapEntry_:
  - Resolve any environment variables specified in either <BuildContextReference> or <BuildContextForImageReference>.
  - Calculate the _ContributingSet_ by resolving and applying <BuildContextReference> to the build context.
    - A file name uniquely identifies an element within the _ContributingSet_.
    - Recursively visit subdirectories, qualifying each file name with the path relative to the root files.  
  - For the resulting _ContributingSet_ assign <BuildContextForImageReference>, as its _ContributingSetLabel_. The _ContributingSetLabel_ itself becomes an element in the set called the <a href="#TermLocalBuildContext">_LocalBuildContext_</a>.  
  - Add or merge the element named _ContributingSetLabel_ to the _LocalBuildContext_.
    - if the _LocalBuildContext_ already contains an element with the same _ContributingSetLabel_, use set [Union](http://en.wikipedia.org/wiki/Union_%28set_theory%29) to merge both sets of _ContributingSet_.
      - In the situation where elements are identically named, perform an additional check to ensure, from a repeatability perspective, identical content and metadata.
        -  If completely identical, discard the duplicate.
        -  Otherwise, generate a build error describing an inability to ensure repeatability.
- The NULL keyword produces an empty LocalBuildContext, prohibiting an <a href="#TermImageContext">_ImageContext_</a>'s Dockerfile commands from accessing resources located in the build context.
- Re-assembly of the <a href="#TermLocalBuildContext">_LocalBuildContext_</a> occurs when the value of any environment variable specified in the references define by _ContextMapEntry_ changes.
- The current caching model, if it considers the file name metadata to include only a file's/directory's name, absent its path within the build context, will be oblivious to the mapping mechanism.
- To remain compatible with existing Dockerfile code, the absence of the CONTEXT keyword permits an _ImageContext_ to directly access the build context.
### Benefits <a name="Benefits"></a>
- Encourages encapsulation from both the perspective of a Dockerfile and an <a href="#TermImageContext">_ImageContext_</a>.  A Dockerfile exposes only the resources in a build context it wishes an _ImageContext_ to access, while an _ImageContext_ can statically, independently define its required interface.  For example, a secret key can be delivered to exactly the required _ImageContext_ without exposing it to any other _ImageContext_ encoded within a multi FROM Dockerfile, greatly reducing the attack surface of the vulnerability mentioned above.
- Promotes encoding of ONBUILD triggers, as these commands can bind to statically defined references that are then dynamically resolved when executed.  ONBUILD triggers encapsulate repetitive code, thus reducing/eliminating manually generated code.  This concision increases a Dockerfile's [declarative benefit](https://github.com/docker/docker/issues/8660#declarativebenefit).
- Adapts the structure of directories/files that exist in other committed containers, exposed by [#12415](https://github.com/docker/docker/issues/12415), to conform to the source directories/file references specified by ADD/COPY.  The ADD/COPY command source references reflect an interface that's oblivious to their actual physical location. Therefore, when the physical location of a directory/file changes, mappings specified by CONTEXT change but the source directory/file references specified by ADD/COPY operators remain the same.
- Enables the partitioning of an aggregated build context - a Dockerfile with more than one FROM statement.
- Since CONTEXT declares the minimal dependencies required by a given FROM, these declared dependencies can be utilized to assist, along with information from [#12415](https://github.com/docker/docker/issues/12415), in determining the execution order and schedule of potentially concurrent build steps (multiple FROMs).
### Example <a name="Example"></a>

Create three different go applications within one Dockerfile using the [google/golang-runtime](https://registry.hub.docker.com/u/google/golang-runtime/dockerfile/) image.  It would be difficult to achieve the result below, with the given build context, without CONTEXT support, as golang-runtime relies on the following trigger "ONBUILD ADD . /gopath/src/app/".  
##### Build Context

```
    /
      Dockerfile
    /app1
      main.go
    /app2
      main.go
    /app3
      main.go
```
##### Dockerfile

```
FROM google/golang-runtime CONTEXT /app1/main.go:/
FROM google/golang-runtime CONTEXT /app2/*:/
FROM google/golang-runtime CONTEXT /app3/:/
```

In each FROM above, the go source code is mapped to the root directory of the <a href="#TermLocalBuildContext">_LocalBuildContext_</a> seen by the trigger statements executing within the <a href="#TermImageContext">_ImageContext_</a> defined by google/golang-runtime.  Therefore, files like the Dockerfile, that would typically be silently consumed and appear in the image derived from google/golang-runtime are excluded.

Signed-off-by: Rich Moyse rich@moyse.us

### CODE REVIEW

**Comment by WhisperingChaos** [comment]
@GordonTheTurtle Sorry, it's signed now.  Thanks!

**Comment by icecrime** [comment]
Sorry, unless I'm missing something, I'm -1 on design:
* The context path can already be provided as part of the `docker build` command argument
* The path of the Dockerfile itself can also be specified with the `docker build -f` comment-line flag
* I like the simplicity that all Dockerfile paths are relative to the context (easy to read, easy to understand)

**Comment by WhisperingChaos** [comment]
@icecrime 

Based on your reply, I believe there's a misconception regarding the issue address by this pull request.  Given the replies: 
>
 + The context path can already be provided as part of the docker build command argument
 + The path of the Dockerfile itself can also be specified with the docker build -f comment-line flag
 + I like the simplicity that all Dockerfile paths are relative to the context (easy to read, easy to understand)

Suggests that the feature described by this pull request is a replacement for or augments the ```docker build``` command line parameter specifying the build context.  It neither replaces nor attempts to extend the build context specified by ```docker build```.  In other words, this feature doesn't affect the external nature of docker build at all.  

**Instead, the proposed feature maps, in a read only way, the directory structure of the uploaded build context provided by ```docker build``` to present a virtual directory structure that corresponds to the one required by a given FROM statement.  This virtual directory structure matches the build context expected by FROM.  Think of it as creating a directory structure comprised entirely of symbolic links that conforms to what's needed by FROM and its associated Dockerfile commands.**

For example, the google/golang-runtime ONBUILD trigger expects go source code to be located in the root directory of the build context provided to it.  As defined by the pull request's example above, the uploaded build context doesn't match this configuration, as it's attempting, within the same Dockerfile to produce three go programs.  Therefore, this aggregated build context must be mapped, in this case partitioned, to reflect the context accepted by google/golang-runtime.  

Borrowing from the example above, the CONTEXT keyword maps the uploaded build context file "/app1/main.go" (not the one that exists on the host)  to the virtual build context expected by google/golang-runtime, that is: "/main.go"  In other words, the following ```FROM google/golang-runtime CONTEXT /app1/main.go:/``` is given its own local build context, implemented as a virtual directory, that simply contains "/main.go" excluding all the other path/file references defined by the uploaded build context.  Therefore, when google/golang-runtime ```ONBUILD ADD . /gopath/src/app``` trigger fires only "main.go" is copied to "/gopath/src/app".

This feature not only benefits existing Dockerfiles, with only a single FROM but Dockerfiles that incorporate multiple FROMs (images) to compose a minimal resulting one.  There are several proposals such as #7115, #7149, #8660, and #12415 that support multiple FROM statements within a Dockerfile that could benefit from the proposed mapping mechanism.

Please consider reopening this pull request.

**Comment by WhisperingChaos** [comment]
@icecrime 

I realize the effort involved in reviewing wave after wave of issues/proposals is time consuming/exhausting, so it's not surprising that my initial request to reopen #12072 wasn't immediately acted on.  However, it's been nearly 1 week since I directed a post to you which explained why I believe there's been a misunderstanding regarding the proposal's intent and why it should be reopened.  I would appreciate a reply to this request.  Thanks.

**Comment by icecrime** [comment]
I think it would help maintainers if you just give us a 2 sentence about what you're trying to solve, and how. I'm reopening this, but you should know that much simpler things stayed opened for much longer: if it can't be broken down, it's very unlikely going to ever see progress.

**Comment by WhisperingChaos** [comment]
@icecrime 

Thanks! Appreciate the advice.  My more recent proposal incorporates a TLDR section.

**Comment by WhisperingChaos** [comment]
Added TLDR and additional Benefits related to #12415.

**Comment by cpuguy83** [comment]
I think this would be better served by allowing the Dockerfile path to be outside of the supplied context... for example `docker build -f ./Dockerfile-runtime app3/`

**Comment by WhisperingChaos** [comment]
@cpuguy83  Thanks for reviewing this proposal.

In response:

+ This proposal permits flexible aggregation and partitioning of the build context.  It is intended to support multiple ```FROM``` statements within a single Dockerfile.  Incorporating many ```FROM``` statements requires a means to initially aggregate the build context with the more elemental build contexts required by each ```FROM``` then partition this aggregate to supply the specific build context for an individual ```FROM```.  Therefore, it's utility increases with each ```FROM``` added to a single Dockerfile.  For example, try incorporating two ```FROM``` statements within a single Dockerfile, one using [google/golang-runtime](https://registry.hub.docker.com/u/google/golang-runtime/dockerfile/) while the other [docker-library/rails] (https://github.com/docker-library/rails/blob/9fb5d2b7e0f2e7029855028e07e86ab7ec54abaa/onbuild/Dockerfile).  Since these ```FROM``` statements use ONBUILD triggers executing ```ADD/COPY . /...``` a single build context cannot currently be crafted to satisfy their needs.
+ The synergy between this proposal and #12415 permits the isolation of build-time and run-time concerns while also extending of the initial build context to include intermediate artifacts.  When used with #12415, this proposal provides a means to create the run-time file system for any arbitrarily complex container using only a **single** ```ADD``` statement (single layer) while at the same time, preserve the highly layered, cached build-time containers.  Please view #12415 and its [Example](https://github.com/docker/docker/issues/12415#Example) to better appreciate this synergy.
+ Although the suggested alternative build request provided by your post will satisfy the last ```FROM``` statement in the example Dockerfile, it cannot meet the requirements of the preceding ```FROM``` statements that appear in it.  Therefore, the alternative solution **does not, nor can it, provide the same outcome** given a single Dockerfile.
+ This proposal provides for restructuring the build context to reflect **any desired** structure.  Therefore, besides enabling the segregation of the Dockerfile from the build context, like the ```-f``` option, its restructuring mechanism can be applied to any other file or directory that's part of the build context – it's more general.  Except for ```-f```'s ability to inform ```docker build``` to use a different Dockerfile name, it would be unnecessary.

**Comment by WhisperingChaos** [comment]
@cpuguy83 
Given the reasoning provided by the [reply](https://github.com/docker/docker/pull/12072#issuecomment-104766304) to your initial post, do you continue to 
> ...think this would be better served by allowing the Dockerfile path to be outside of the supplied context... for example docker build -f ./Dockerfile-runtime app3/

If you still think this way, would you provide your reasoning or identify misconceptions in my statements?

**Comment by icecrime** [comment]
Sorry but it's been two months, and I don't see neither user nor maintainer enthusiasm on this :-/ I can't justify keeping this open. Feel free to advocate for this on one of our groups (such as docker-dev), we are obviously open to changing our minds if there's obvious and significant interest for it. Thanks!

**Comment by WhisperingChaos** [comment]
@icecrime 

I appreciate your notice to close this pull request and suggestion to continue advocating for its inclusion via docker-dev, instead of simply closing it.  Take Care!
